README.md
# carousel_3d

A new Flutter package.

## TransformerPageView简易的3D翻转效果

   Carousel3DWidget
   
##Installation
  carousel_3d:
##simple
```dart
import 'package:carousel_3d/carousel_3d.dart';
   
    ...
    
  Carousel3DWidget(autoPlay: false, //自动播放
         loop: true,//无限循环
         duration: 1000,//反循环效果时间
         viewportFraction: 0.6,//所占宽度比例
         scale: 0.7,//翻转放缩比例
         itemBuilder: (BuildContext context, int index) {
               return Image.asset(imgs[index],fit: BoxFit.cover,);
         },
         itemCount: imgs.length,)
```
  

CHANGELOG.md
## [0.0.1] - TODO: Add release date.

* TODO: Describe initial release.
* Carousel3DWidget 新建

## [0.0.2]

* 清理日志，优化效果

## [0.0.3]
* 更新readme

lib/src/carousel_3d.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:transformer_page_view/index_controller.dart';
import 'package:transformer_page_view/transformer_page_view.dart';

import 'transformer_3d.dart';


class Carousel3DWidget extends StatefulWidget {
  bool loop;
  int duration;
  ValueChanged<int> onChanged;
  int index;
  int itemCount;
  double viewportFraction;
  IndexedWidgetBuilder itemBuilder;
  bool autoPlay;
  int autoPlayDelay;
  double scale;

  Carousel3DWidget(
      {Key key,
      this.loop = true,
      this.duration = 1000,
      this.onChanged,
      this.index = 0,
      this.itemCount,
      this.viewportFraction = 0.6,
      this.itemBuilder,
      this.autoPlay = true,
      this.autoPlayDelay = 1500,
      this.scale = 0.7})
      : assert(itemBuilder != null || itemCount != null,
            "itemBuilder and transformItemBuilder must not be both null"),
        super(key: key);

  @override
  _Carousel3DWidgetState createState() => _Carousel3DWidgetState();
}

class _Carousel3DWidgetState extends State<Carousel3DWidget> {
  Timer _timer;
  IndexController _controller;

  @override
  void initState() {
    super.initState();
    _controller = IndexController();
    if (widget.autoPlay) {
      _startAutoPlay();
    }
  }

  @override
  Widget build(BuildContext context) {
    return NotificationListener(
      child: TransformerPageView(
        controller: _controller,
        key: UniqueKey(),
        loop: widget.loop,
        duration: Duration(milliseconds: widget.duration),
        onPageChanged: (int i) {
          if (widget.onChanged != null) widget.onChanged(i);
        },
        pageController: TransformerPageController(
            initialPage: widget.index,
            loop: widget.loop,
            itemCount: widget.itemCount,
            reverse: true,
            viewportFraction: widget.viewportFraction),
        transformer: Transformer3D(context,
            scale: widget.scale, viewportFraction: widget.viewportFraction),
        // index: bean.tag,
        itemBuilder: (BuildContext context, int i) {
          return widget.itemBuilder(context, i);
        },
        itemCount: widget.itemCount,
        viewportFraction: widget.viewportFraction ?? 0.6,
      ),
      onNotification: (ScrollNotification notification) {
        if (notification is ScrollStartNotification) {
          if (notification.dragDetails != null) {
            //by human
            if (_timer != null) _stopAutoPlay();
          }
        } else if (notification is ScrollEndNotification) {
          if (_timer == null) _startAutoPlay();
        }

        return false;
      },
    );
  }

  void _startAutoPlay() {
    if (!widget.autoPlay) return;
    assert(_timer == null, "Timer must be stopped before start!");
    _timer =
        Timer.periodic(Duration(milliseconds: widget.autoPlayDelay), _onTimer);
  }

  void _onTimer(Timer timer) {
    _controller.next(animation: true);
  }

  void _stopAutoPlay() {
    if (_timer != null) {
      _timer.cancel();
      _timer = null;
    }
  }
}


lib/src/transformer_3d.dart
import 'package:flutter/src/widgets/framework.dart';
import 'package:flutter/widgets.dart';
import 'package:transformer_page_view/transformer_page_view.dart';

class Transformer3D extends PageTransformer {
  final double scale;
  final double fade;
  final double viewportFraction;
  BuildContext context;

  Transformer3D(this.context,
      {this.fade = 0.5, this.scale = 0.7, this.viewportFraction});

  @override
  Widget transform(Widget item, TransformInfo info) {
    double position = info.position;
    Widget child = item;
    double scaleF = 0;
    if (scale != null) {
      double scaleFactor = (1 - position.abs()) * (1 - scale);
      scaleF = scale + scaleFactor;

      child = new Transform.scale(
        scale: scaleF,
        child: item,
      );
    }
    if (fade != null) {
      double fadeFactor = (1 - position.abs()) * (1 - fade);
      double opacity = fade + fadeFactor;
      child = new Opacity(
        opacity: opacity,
        child: child,
      );
    }

    // if (position.abs() != 0) {
    //   child = ClipRect(
    //     clipper: PageCustomClipper((-position * position.abs()) * info.width,
    //         info.width * viewportFraction),
    //     child: child,
    //   );
    // }

    child = Transform.translate(
      offset: Offset((position * position.abs()) * info.width, 0),
      child: child,
    );
    child = ClipRect(
      child: child,
    );
    return child;
  }
}

class PageCustomClipper extends CustomClipper<Rect> {
  double left;
  double width;

  PageCustomClipper(this.left, this.width);

  @override
  bool shouldReclip(covariant CustomClipper<Rect> oldClipper) {
    return true;
  }

  @override
  Rect getClip(Size size) {
    return Rect.fromLTWH(left, 0, width, size.height);
  }
}


example/lib/main.dart
import 'package:carousel_3d/carousel_3d.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  var imgs = [
    'rescoucs/img/img_1.jpeg',
    'rescoucs/img/img_2.jpeg',
    'rescoucs/img/img_3.webp'
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Container(
              width: 350,
              height: 350,
              child: Carousel3DWidget(
                autoPlay: false,
                loop: true,
                duration: 1000,
                viewportFraction: 0.6,
                scale: 0.7,
                itemBuilder: (BuildContext context, int index) {
                  return Image.asset(
                    imgs[index],
                    fit: BoxFit.cover,
                  );
                },
                itemCount: imgs.length,
              ),
            ),
          ],
        ),
      ),
    );
  }
}
